Sprint 10

pavelseleznev opened the pull request:
Implemented OAuth2.0 authorization/network stack, added login screen.

So1ovyov commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,55 @@
+        vc.dismiss(animated: true)
+        
+        oauth2Service.fetchOAuthToken(code: code) { [weak self] result in
+            guard let self = self else { return }
```
Подобные конструкции можно делать немного короче:
```
guard let self else { return }
```

So1ovyov commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,55 @@
+
+final class AuthViewController: UIViewController {
+    
+    // MARK: - Internal Properties
```
Кажется, что ниже приватные свойства, поэтому логичнее назвать МАРК:
```
// MARK: - Private Properties
```

So1ovyov commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,55 @@
+            guard let self = self else { return }
+            switch result {
+            case .success(let accessToken):
+                let tokenStorage = OAuth2TokenStorage()
```
Предлагаю создать один раз экземпляр хранилища, а не в каждом новом вызове авторизации:
```
final class AuthViewController: UIViewController {
    
    // MARK: - Private Properties

    private let tokenStorage = OAuth2TokenStorage()
```
И использовать его внутри `success` кейса:
```
        oauth2Service.fetchOAuthToken(code: code) { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let accessToken):
                tokenStorage.token = accessToken
```

So1ovyov commented on ImageFeed/Auth/URLSession+data.swift:
```diff
@@ -0,0 +1,69 @@
+
+import Foundation
+
+enum NetworkError: Error {
```
Давай придерживаться принципа 1 файл = 1 структура/класс/модель/enum. Не стоит раздувать файл и располагать в нем все сразу.
Можно создать директорию `Models` и в нее вынести файлы `NetworkError.swift` и `HTTPStatusCode.swift` которые будут отдельно хранить свои сущности.

So1ovyov commented on ImageFeed/Auth/URLSession+data.swift:
```diff
@@ -0,0 +1,69 @@
+            }
+        }
+        
+        Task {
```
Какое-то смешение `GCD` и `Swift Concurrency`, нет ли желания сделать в едином стиле ?
Например:
```
extension URLSession {
    func data(for request: URLRequest, completion: @escaping (Result<Data, Error>) -> Void){
        let fulfillCompletionOnTheMainThread: (Result<Data, Error>) -> Void = { result in
            DispatchQueue.main.async {
                completion(result)
            }
        }

        dataTask(with: request) { data, response, error in
            if let data = data, let response = response, let statusCode = (response as? HTTPURLResponse)?.statusCode {
                if 200 ..< 300 ~= statusCode {
                    fulfillCompletionOnTheMainThread(.success(data))
                } else {
                    let handleError = self.showNetworkError(for: statusCode)
                    fulfillCompletionOnTheMainThread(.failure(handleError))
                }
            } else if let error = error {
                fulfillCompletionOnTheMainThread(.failure(NetworkError.internalServerError))
            } else {
                fulfillCompletionOnTheMainThread(.failure(NetworkError.internalServerError))
            }
        }
    }

    private func showNetworkError(for statusCode: Int) -> NetworkError {
        if let httpsCode = HTTPStatusCode(rawValue: statusCode) {
            switch httpsCode {
            case .authorizationRequiredStatus:
                return NetworkError.authorizationRequiredError
            case .webPageNotFoundStatus:
                return NetworkError.webPageNotFoundError
            case .internalServerErrorStatus:
                return NetworkError.internalServerError
            }
        }
        return NetworkError.urlSessionError
    }
}
```

pavelseleznev replied:
Вот это изменение тоже хотел бы внести. С данным кодом после авторизации моё приложение не перекидывает на TabBarViewController (где список фотографий).

Потребуется время чтобы разобраться.

pavelseleznev replied:
Данное изменение также внёс. Ещё раз спасибо.

So1ovyov commented on ImageFeed/Auth/URLSession+data.swift:
```diff
@@ -0,0 +1,69 @@
+    }
+    
+    /// Checks for response code from the server
+    func loadNetworkData(request: URLRequest) async throws -> Data {
```
Этот и следующий методы используются только внутри `func data(for request` и их можно заприватить

So1ovyov commented on ImageFeed/Auth/URLSession+data.swift:
```diff
@@ -0,0 +1,69 @@
+                let data = try await self.loadNetworkData(request: request)
+                fulfillCompletionOnTheMainThread(.success(data))
+            } catch {
+                print("Failed to loadNetworkData request")
```
В данном случае не понятно какая ошибка произошла, давай добавим в `print` код ошибки или ее название.

So1ovyov commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,127 @@
+    /// Loads the login page using Unsplash web address. Provides the required data - access key, redirect uri etc.
+    private func loadAuthView() {
+        guard var urlComponents = URLComponents(string: Constants.authorizeURLString) else
+        { return }
```
```
guard var urlComponents = URLComponents(string: Constants.authorizeURLString) else { return }
/// или
        guard var urlComponents = URLComponents(string: Constants.authorizeURLString) else {
            return
        }
```

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,70 @@
+
+import Foundation
+
+struct OAuthTokenResponseBody: Decodable {
```
Давай это также вынесем в отдельный файл.

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -0,0 +1,70 @@
+        
+        return request
+    }
+}
```
Тут все отлично!

So1ovyov commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2TokenStorage.swift:
```diff
@@ -0,0 +1,34 @@
+        }
+        set {
+            guard let newValue else {
+                assertionFailure("Invalid token")
```
Использование выражений не приветствуется внутри основного кода приложения, лучше только в тестах:
https://stevenpcurtis.medium.com/precondition-assert-fatal-error-or-guard-in-your-swift-code-5f9297658be0
```
    var token: String? {
        get { userDefaults.string(forKey: Keys.token.rawValue) }
        set {
            guard let newValue else { return }
            userDefaults.set(newValue, forKey: Keys.token.rawValue)
        }
    }
```

So1ovyov commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,78 @@
+    
+    // MARK: - Private Properties
+    /// Identifier to show initial authentication screen
+    private var showAuthenticationScreenSegueIdentifier = "showAuthenticationScreen"
```
Тут сразу несколько моментов:
```
final class SplashViewController: UIViewController {
    
    // MARK: - Private Properties

    private let showAuthenticationScreenSegueIdentifier = "showAuthenticationScreen"
    private let oauth2TokenStorage: OAuth2TokenStorageProtocol = OAuth2TokenStorage()
```
1. Предлагаю комментарии для каждого свойства удалить, по коду все и так предельно понятно, а описание каждого свойства визуально увеличивает код
2. У свойства `showAuthenticationScreenSegueIdentifier` достаточно константы, а не переменной

So1ovyov commented:
@pavelseleznev отличная работа! Осталось поправить дизайн пары элементов и если будет возможность применить рекомендации по коду. Жду исправлений.

pavelseleznev commented:
Applied critical/non-critical post review fixes:
1) Changed the background color scheme of the project to YP_black_color for consistent and uniform look. Corrected the font for the login button. Fixed the logo size on the launch screen.
2) Compartmentalized enums/protocols into different files.
3) Replaced assertionFailure error print.
4) Minor code/stylistic changes.

pavelseleznev commented:
> @pavelseleznev отличная работа! Осталось поправить дизайн пары элементов и если будет возможность применить рекомендации по коду. Жду исправлений.

Спасибо большое за ревью. Эти замечания ценны для меня.

So1ovyov commented:
> > @pavelseleznev отличная работа! Осталось поправить дизайн пары элементов и если будет возможность применить рекомендации по коду. Жду исправлений.
> 
> Спасибо большое за ревью. Эти замечания ценны для меня.

Я очень рад, что мы рекомендации привнесли что-то новое и полезное в твое обучение.
Также приятно видеть, как студент применяет советы и делает свой код лучше.
Успехов в следующем спринте :3
